src/libaktualizr/primary/aktualizr.cc

```diff
@@ -43,33 +43,54 @@ void Aktualizr::systemSetup() {
   LOG_TRACE << "... seeding complete in " << timer;
 }
 
-void Aktualizr::Initialize() { uptane_client_->initialize(); }
-
-void Aktualizr::UptaneCycle() {
-  RunningMode running_mode = config_.uptane.running_mode;
-  UpdateCheckResult update_result = CheckUpdates();
-  if (running_mode == RunningMode::kCheck || update_result.updates.size() == 0) {
-    return;
-  } else if (running_mode == RunningMode::kInstall) {
-    Install(update_result.updates);
-    uptane_client_->putManifest();
-    return;
-  }
+void Aktualizr::Initialize() {
+  std::thread([&]() {
+    while (!shutdown_) {
+      auto task = api_queue_.dequeue();
+      if (shutdown_) {
+        return;
+      }
+      task();
+    }
+  })
+      .detach();
+  uptane_client_->initialize();
+}
 
-  DownloadResult download_result = Download(update_result.updates);
-  if (running_mode == RunningMode::kDownload || download_result.status != DownloadStatus::kSuccess ||
-      download_result.updates.size() == 0) {
-    return;
-  }
+std::future<void> Aktualizr::UptaneCycle() {
+  auto promise = std::make_shared<std::promise<void>>();
+  std::function<void()> task([&, promise]() {
+    RunningMode running_mode = config_.uptane.running_mode;
+    UpdateCheckResult update_result = CheckUpdates().get();
+    if (running_mode == RunningMode::kCheck || update_result.updates.size() == 0) {
+      promise->set_value();
+      return;
+    } else if (running_mode == RunningMode::kInstall) {
+      Install(update_result.updates);
+      uptane_client_->putManifest();
+      promise->set_value();
+      return;
+    }
+
+    DownloadResult download_result = Download(update_result.updates).get();
+    if (running_mode == RunningMode::kDownload || download_result.status != DownloadStatus::kSuccess ||
+        download_result.updates.size() == 0) {
+      promise->set_value();
+      return;
+    }
 
-  Install(download_result.updates);
-  uptane_client_->putManifest();
+    Install(download_result.updates);
+    uptane_client_->putManifest();
+    promise->set_value();
+  });
+  api_queue_.enqueue(task);
+  return promise->get_future();
 }
 
 int Aktualizr::RunForever() {
-  SendDeviceData();
+  SendDeviceData().get();
   while (!shutdown_) {
-    UptaneCycle();
+    UptaneCycle().get();
     std::this_thread::sleep_for(std::chrono::seconds(config_.uptane.polling_sec));
   }
   return EXIT_SUCCESS;
```

Comment:
While `RunForever` does not need to return a future, it probably should be asynchronous and return immediately.

Reply:
Do you mean it should run in separate thread?

Reply:
Yeah, mostly likely. Maybe there's a clever way to reuse the API queue thread, maybe not.

Reply:
Yes, we cannot, because we will get dead lock.